crates/client/rpc/src/trace_api.rs

```diff
@@ -244,6 +244,164 @@ where
 
         Ok(traces)
     }
+
+    async fn trace_transaction(&self, transaction_hash: FieldElement) -> RpcResult<TransactionTraceWithHash> {
+        let substrate_block_hash = self
+            .backend
+            .mapping()
+            .block_hash_from_transaction_hash(Felt252Wrapper(transaction_hash).into())
+            .map_err(|e| {
+                error!("Failed to get transaction's substrate block hash from mapping_db: {e}");
+                StarknetRpcApiError::TxnHashNotFound
+            })?
+            .ok_or(StarknetRpcApiError::TxnHashNotFound)?;
+
+        let starknet_block = get_block_by_block_hash(self.client.as_ref(), substrate_block_hash)?;
+        let chain_id = Felt252Wrapper(self.chain_id()?.0);
+
+        let block_until_tx = starknet_block
+            .transactions()
+            .iter()
+            .take_while(|tx| {
+                let tx_hash = tx.compute_hash::<H>(chain_id, false);
+                tx_hash != transaction_hash.into()
+            })
+            .map(|tx| match tx {
```

Comment:
This whole map is a copy paste of ` trace_block_transactions`, factorize the code, make it an independent function